File: Infinite Scroll (ObjC)/InfiniteScrollSample/ArticleListViewController.m

#import "ArticleListViewController.h"
#import "ArticleCell.h"
#import "SponsoredLandingPageViewController.h"
#import "AppDelegate.h"
#import "ArticleViewController.h"
#import <WebKit/WebKit.h>
#import <UIKit/UIKit.h>


// Nativo Settings
static NSString * const ArticleCellIdentifier = @"articlecell";
static NSString * const VideoCellIdentifier = @"videocell";
static NSString * const NativoSectionUrl = @"http://www.publisher.com/test";

// Define the frequency of Ad cells for infinite scroll
#define AD_ROW_START 4
#define AD_ROW_INTERVAL 10

@interface ArticleListViewController () <UITableViewDataSource, UITableViewDelegate, NtvSectionDelegate>
@property (nonatomic) NSCache *feedImgCache;
@property (nonatomic) NSDateFormatter *dateFormatter;
@property (nonatomic) IBOutlet UITableView *tableView;
@property (nonatomic) NSMutableArray *articlesDataSource;
@end

@implementation ArticleListViewController


#pragma mark - Initialization methods

- (void)viewDidLoad
{
    [super viewDidLoad];
    [self setupView];
    
    // Enable Test Nativo Test & Debug mode
    [NativoSDK enableDevLogs];
    [NativoSDK enableTestAdvertisements];

    // Setup Section
    [NativoSDK setSectionDelegate:self forSection:NativoSectionUrl];
    
    // Register Nativo template types using Nib files
    [NativoSDK registerNib:[UINib nibWithNibName:@"ArticleNativeAdView" bundle:nil] forAdTemplateType:NtvAdTemplateTypeNative];
    [NativoSDK registerNib:[UINib nibWithNibName:@"ArticleVideoAdView" bundle:nil] forAdTemplateType:NtvAdTemplateTypeVideo];
    
    [NativoSDK registerNib:[UINib nibWithNibName:@"SponsoredLandingPageViewController" bundle:nil] forAdTemplateType:NtvAdTemplateTypeLandingPage];
    
    // Cell used as nativo ad container
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"nativocell"];
        
    [self startArticleFeed];
}


#pragma mark - NtvSectionDelegate Methods

- (void)section:(NSString *)sectionUrl needsPlaceAdInViewAtLocation:(id)identifier {
    [self.tableView reloadData];
}

- (void)section:(NSString *)sectionUrl needsRemoveAdViewAtLocation:(id)identifier {
    [self.tableView reloadData];
}

- (void)section:(NSString *)sectionUrl needsDisplayLandingPage:(nullable UIViewController *)sponsoredLandingPageViewController {
    //Push the sponsored landing page to the navigationController
    [self.navigationController pushViewController:sponsoredLandingPageViewController animated:YES];
}

- (void)section:(NSString *)sectionUrl needsDisplayClickoutURL:(NSURL *)url {
    
    // Load click-out url in WKWebView
    UIViewController *clickoutAdVC = [[UIViewController alloc] init];
    WKWebView *webView = [[WKWebView alloc] initWithFrame:clickoutAdVC.view.frame];
    [clickoutAdVC.view addSubview:webView];
    NSURLRequest *clickoutReq = [NSURLRequest requestWithURL:url];
    [webView loadRequest: clickoutReq];
    [self.navigationController pushViewController:clickoutAdVC animated:YES];
}

#pragma mark - UITableViewDataSource methods

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (indexPath.row == self.articlesDataSource.count-1) {
        [self getNextFeedItems];
    }
    
    UITableViewCell *cell = nil;
    BOOL isNativoAdPlacement = indexPath.row % AD_ROW_INTERVAL == AD_ROW_START;
    BOOL didGetNativoAdFill = NO;
    if (isNativoAdPlacement) {
        cell = [tableView dequeueReusableCellWithIdentifier:@"nativocell"];
        didGetNativoAdFill = [NativoSDK placeAdInView:cell atLocationIdentifier:indexPath inContainer:tableView forSection:NativoSectionUrl options:nil];
    }
    
    // If not a Nativo ad placement, or if the placement had no fill, we must populate the cell with article data
    if (!didGetNativoAdFill) {
        cell = [tableView dequeueReusableCellWithIdentifier:ArticleCellIdentifier];
        
        // One last step. Here we ask the NativoSDK to adjust the indexpath to account for any ads we may have recieved.
        NSIndexPath *adjustedIndexPathWithAds = [NativoSDK getAdjustedIndexPath:indexPath forAdsInjectedInSection:NativoSectionUrl];
        NSDictionary *feedItem = self.articlesDataSource[adjustedIndexPathWithAds.row];
        
        // Populate cell with data
        [self populateCell:(ArticleCell *)cell WithData:feedItem];
    }

    return cell;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    NSInteger numAds = [NativoSDK getNumberOfAdsInSection:NativoSectionUrl inTableOrCollectionSection:section forNumberOfItemsInDatasource:self.articlesDataSource.count];
    return self.articlesDataSource.count + numAds;
}



#pragma mark - UITableViewDelegate methods

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Ask the NativoSDK for adjusted indexpath in order to get the correct data for the row.
    NSIndexPath *adAdjustedIndexPath = [NativoSDK getAdjustedIndexPath:indexPath forAdsInjectedInSection:NativoSectionUrl];
    
    NSDictionary *feedItem = [self.articlesDataSource objectAtIndex:adAdjustedIndexPath.row];
    ArticleViewController *article = [[ArticleViewController alloc] initWithNibName:@"ArticleViewController" bundle:nil];
    NSString *articleUrlStr = [NSString stringWithFormat:@"https://www.nativo.com%@", feedItem[@"fullUrl"]];
    article.articleURL = [NSURL URLWithString:articleUrlStr];
    [self.navigationController pushViewController:article animated:YES];
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}


#pragma mark - Feed Request

- (void)startArticleFeed {
    NSString *filePath = [[NSBundle mainBundle] pathForResource:@"nativoblog" ofType:@"json"];
    NSData *feedData = [NSData dataWithContentsOfFile:filePath options:0 error:nil];
    NSDictionary *feed = [NSJSONSerialization JSONObjectWithData:feedData options:0 error:nil];
    NSArray *feedItems = feed[@"items"];
    [self.articlesDataSource addObjectsFromArray:feedItems];
}

- (void)getNextFeedItems {
    [self.articlesDataSource addObjectsFromArray:[self.articlesDataSource copy]];
    [self.tableView reloadData];
}


#pragma mark - Misc

- (void)populateCell:(ArticleCell *)cell WithData:(NSDictionary *)feedItem {
    cell.titleLabel.text = feedItem[@"title"];
    cell.dateLabel.text = [self.dateFormatter stringFromDate:[NSDate date]];
    cell.authorNameLabel.text = feedItem[@"author"];
    cell.previewTextLabel.text = feedItem[@"excerpt"];
    
    // async fetch image
    NSString *imgStr = feedItem[@"assetUrl"];
    [self getAsyncImageForUrl:[NSURL URLWithString:imgStr] completion:^(UIImage *img) {
        if (img && [cell.titleLabel.text isEqualToString:feedItem[@"title"]]) {
            cell.adImageView.image = img;
        }
    }];
}

- (void)getAsyncImageForUrl:(NSURL *)url completion:(void (^)(UIImage *img))block
{
    dispatch_async(dispatch_get_global_queue(0, 0), ^{
        if (!self.feedImgCache) {
            self.feedImgCache = [[NSCache alloc] init];
        }
        UIImage *img = [self.feedImgCache objectForKey:url];
        if (img) {
            dispatch_sync(dispatch_get_main_queue(), ^{
                block(img);
            });
            return;
        }
        NSData *imgData = [NSData dataWithContentsOfURL:url];
        UIImage *image = [UIImage imageWithData:imgData];
        if (image) {
            [self.feedImgCache setObject:image forKey:url]; // Set image on cache
        }
        dispatch_sync(dispatch_get_main_queue(), ^{
            block(image);
        });
    });
}

- (void) setupView {
    self.articlesDataSource = [NSMutableArray array];
    UIView *footerView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.tableView.bounds.size.width, 50.0f)];
    UIActivityIndicatorView *loadingIndicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleGray];
    loadingIndicator.center = footerView.center;
    [footerView addSubview:loadingIndicator];
    [loadingIndicator startAnimating];
    [self.tableView setTableFooterView:footerView];
}

- (NSDateFormatter *)dateFormatter
{
    if (!_dateFormatter) {
        // Create date formatter for later
        _dateFormatter = [[NSDateFormatter alloc] init];
        [_dateFormatter setDateStyle:NSDateFormatterShortStyle];
        [_dateFormatter setTimeStyle:NSDateFormatterShortStyle];
    }
    return _dateFormatter;
}

@end

File: Table View (Swift)/NativoSwiftTableViewSample/ArticleListViewController.swift
//
//  ViewController.swift
//  NativoSwiftTableViewSample
//
//  Copyright © 2019 Nativo. All rights reserved.
//

import UIKit
import NativoSDK

class ArticleListViewController: UIViewController {

    @IBOutlet weak var tableView: UITableView!
    var articlesDataSource  = Array<Dictionary<String, Any>>()
    var feedImgCache = Dictionary<URL, UIImage>()
    let dateFormatter = DateFormatter()
    let ArticleCellIdentifier = "articlecell"
    let VideoCellIdentifier = "videocell"
    let NativoSectionUrl = "http://www.publisher.com/test"
    let NativoAdRow1 = 3
    let NativoAdRow2 = 9
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupView()
        
        NativoSDK.enableDevLogs()
        NativoSDK.enableTestAdvertisements(with: NtvTestAdType.native)
        
        NativoSDK.setSectionDelegate(self, forSection: NativoSectionUrl)
        NativoSDK.registerReuseId(ArticleCellIdentifier, for: .native) // These identifiers correlate to the dynamic prototype cells set in Main.storyboard
        NativoSDK.registerReuseId(VideoCellIdentifier, for: .video)
        NativoSDK.register(UINib(nibName: "SponsoredLandingPageViewController", bundle: nil), for: .landingPage)
        
        // Register blank cell to be used as container for Nativo ads
        self.tableView.register(UITableViewCell.classForCoder(), forCellReuseIdentifier: "nativocell")
        self.tableView.prefetchDataSource = self;
        
        startArticleFeed()
    }
}

extension ArticleListViewController: NtvSectionDelegate {
    
    func section(_ sectionUrl: String, needsReloadDatasourceAtLocationIdentifier identifier: Any, forReason reason: String) {
        self.tableView.reloadData()
    }
    
    func section(_ sectionUrl: String, needsDisplayLandingPage sponsoredLandingPageViewController: (UIViewController & NtvLandingPageInterface)?) {
        if let landingPage = sponsoredLandingPageViewController {
            self.navigationController?.pushViewController(landingPage, animated: true)
        }
    }
    
    func section(_ sectionUrl: String, needsDisplayClickoutURL url: URL) {
        let clickoutAdVC = UIViewController()
        let webView = WKWebView(frame: clickoutAdVC.view.frame)
        let clickoutReq = URLRequest(url: url)
        webView.load(clickoutReq)
        self.navigationController?.pushViewController(clickoutAdVC, animated: true)
        clickoutAdVC.view.addSubview(webView)
    }
}


extension ArticleListViewController: UITableViewDataSource, UITableViewDelegate {
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let numAds = NativoSDK.getNumberOfAds(inSection: NativoSectionUrl, inTableOrCollectionSection: section, forNumberOfItemsInDatasource: self.articlesDataSource.count)
        let totalRows = self.articlesDataSource.count + numAds
        print("-- Total Rows: \(totalRows)")
        return totalRows
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let isNativoAdPlacement: Bool = indexPath.row == NativoAdRow1 || indexPath.row == NativoAdRow2
        var didGetNativoAdFill: Bool = false
        var cell: UITableViewCell! // Will always be initialized in this control flow so we can safely declare as implicitley unwrapped optional
        if isNativoAdPlacement {
            let nativoReuseId = "nativocell" // use seperate reuse id for Nativo ads
            
            // Using NativoSDK View API to inject ad into cell. The response will let you know if ad placement was filled.
            cell = tableView.dequeueReusableCell(withIdentifier: nativoReuseId, for: indexPath)
            didGetNativoAdFill = NativoSDK.placeAd(in: cell, atLocationIdentifier: indexPath, inContainer: tableView, forSection: NativoSectionUrl)
        }
        
        if !didGetNativoAdFill {
            let articleCell: ArticleCell = tableView.dequeueReusableCell(withIdentifier: ArticleCellIdentifier, for: indexPath) as! ArticleCell
            
            // Get the adjusted index path so we account for possible ad displacement in datasource
            let custom : IndexPath = NativoSDK.getAdjustedIndexPath(indexPath, forAdsInjectedInSection: NativoSectionUrl)
            let row = custom.row
            let data = self.articlesDataSource[row]
            self.injectCell(articleCell, withData: data)
            cell = articleCell
        }
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let adAdjustedIndexPath : IndexPath = NativoSDK.getAdjustedIndexPath(indexPath, forAdsInjectedInSection: NativoSectionUrl)
        if adAdjustedIndexPath.row < self.articlesDataSource.count {
            let articleItem = self.articlesDataSource[adAdjustedIndexPath.row]
            let articleUrlStr = "https://www.nativo.com\(articleItem["fullUrl"]!)"
            let articleViewController = ArticleViewController(nibName: "ArticleViewController", bundle: nil)
            articleViewController.articleURL = URL(string: articleUrlStr)
            self.navigationController?.pushViewController(articleViewController, animated: true)
        }
    }
}

extension ArticleListViewController: UITableViewDataSourcePrefetching {
    
    func tableView(_ tableView: UITableView, prefetchRowsAt indexPaths: [IndexPath]) {
        // Prefetch Nativo SDK ads at specified index paths
        for indexPath in indexPaths {
            let isNativoAdPlacement = indexPath.row == NativoAdRow1 || indexPath.row == NativoAdRow2
            if isNativoAdPlacement {
                NativoSDK.prefetchAd(forSection: NativoSectionUrl, atLocationIdentifier: indexPath, options: nil)
            }
        }
    }
}

extension ArticleListViewController {
    
    func setupView() {
        self.dateFormatter.dateStyle = .short
        self.dateFormatter.timeStyle = .short
    }
    
    func startArticleFeed() {
        let filePath = Bundle.main.path(forResource: "nativoblog", ofType: "json")
        let feedData = try! Data.init(contentsOf: URL(fileURLWithPath: filePath!))
        let feed = try! JSONSerialization.jsonObject(with: feedData) as! Dictionary<String, Any>
        let feedItems = feed["items"] as! Array<Dictionary<String, Any>>
        self.articlesDataSource.append(contentsOf: feedItems)
    }
    
    func injectCell(_ cell: ArticleCell, withData data: Dictionary<String, Any> ) {
        cell.titleLabel.text = data["title"] as? String
        cell.authorNameLabel.text = data["author"] as? String
        cell.previewTextLabel.text = data["excerpt"] as? String
        cell.dateLabel.text = DateFormatter.localizedString(from: Date.init(), dateStyle: .medium, timeStyle: .short)
        let imgUrl = data["assetUrl"] as! String
        self.getAsyncImage(forUrl: URL.init(string: imgUrl)!) { (img) in
            if let articleImg = img {
                if cell.titleLabel.text == data["title"] as? String {
                    cell.adImageView.image = articleImg
                }
            }
        }
    }
    
    func getAsyncImage(forUrl url: URL, completion: @escaping (UIImage?) -> Void) {
        DispatchQueue.global(qos: .default).async {
            if let image = self.feedImgCache[url]  {
                DispatchQueue.main.async {
                    completion(image)
                }
                return
            }
            if let imgData = try? Data.init(contentsOf: url) {
                let image = UIImage.init(data: imgData)
//                if (image != nil) {
//                    self.feedImgCache[url] = image
//                }
                DispatchQueue.main.async {
                    completion(image)
                }
            }
        }
    }
}

